nixos/modules/virtualisation/qemu-vm.nix

```diff
@@ -198,6 +198,30 @@ let
         fi
       ''}
 
+      ${lib.optionalString cfg.tpm.enable ''
+        NIX_SWTPM_DIR=$(readlink -f "''${NIX_SWTPM_DIR:-${config.system.name}-swtpm}")
+        mkdir -p "$NIX_SWTPM_DIR"
+        ${lib.getExe cfg.tpm.package} \
+          socket \
+          --tpmstate dir="$NIX_SWTPM_DIR" \
+          --ctrl type=unixio,path="$NIX_SWTPM_DIR"/socket,terminate \
+          --pid file="$NIX_SWTPM_DIR"/pid --daemon \
+          --tpm2 \
+          --log file="$NIX_SWTPM_DIR"/stdout,level=6
+
+        # leave a dangling subprocess
+        # because the swtpm ctrl socket has "terminate" when the last
+        # connection disconnects, it stops swtpm.
+        # When qemu stop, or if the main shell process ends, the coproc
+        # will get signaled by virtue of the pipe between main and coproc
+        # ending. Which in turns triggers a socat connect-disconnect to swtpm
+        # which will stops it.
+        coproc waitingswtpm {
+          read || :
+          echo "" | ${lib.getExe hostPkgs.socat} STDIO UNIX-CONNECT:"$NIX_SWTPM_DIR"/socket
```

Comment:
```suggestion
         # Bash will set the `FD_CLOEXEC` flag on the pipe from parent to coproc
         # this will trigger the coproc to read EOF when qemu is `exec` below.
         # To work around that, we will wait a second before triggering swtpm to exit, assuming qemu will have
         # already connected to swtpm by then.
          sleep 1
          echo "" | ${lib.getExe hostPkgs.socat} STDIO UNIX-CONNECT:"$NIX_SWTPM_DIR"/socket
```

Got to re-read an strace of bash behavior in trying to reply to @ElvishJerricco (https://github.com/NixOS/nixpkgs/pull/246507#issuecomment-1773859259), discovered I missed the `fcntl(63, F_SETFD, FD_CLOEXEC) = 0` the parent runs.

<details>
<summary>
strace output // analysis
</summary>

``` bash
# foo.sh
coproc waitparent {
   read || :
   echo "bar"
}

sleep 1

echo "foo"

exec sleep 2
```

strace output:
```
$ strace -t -f -e clone,dup2,pipe2,write,read,close,fcntl bash foo.sh
[...]
<<< Parent creates a pipe and rebinds parts to FD 63 & 62 // coproc -> parent pipe >>>
21:45:37 pipe2([3, 4], 0)               = 0
21:45:37 dup2(3, 63)                    = 63
21:45:37 close(3)                       = 0
21:45:37 dup2(4, 62)                    = 62
21:45:37 close(4)                       = 0
<<< Parent creates a pipe and rebinds parts to FD 61 & 6 // parent -> coproc pipe >>>
21:45:37 pipe2([3, 4], 0)               = 0
21:45:37 dup2(3, 61)                    = 61
21:45:37 close(3)                       = 0
21:45:37 dup2(4, 60)                    = 60
21:45:37 close(4)                       = 0
<<< Parent spawns the coproc >>>
21:45:37 clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f72dd92fa10) = 1849390
strace: Process 1849390 attached
<<< Parent closes the write part of the coproc -> parent pipe >>>
[pid 1849389] 21:45:37 close(62)        = 0
<<< Parent closes the read part of the parent -> coproc pipe >>>
[pid 1849389] 21:45:37 close(61)        = 0
<<< Parent sets CLOEXEC on its FDs for the pipe >>>
[pid 1849389] 21:45:37 fcntl(63, F_SETFD, FD_CLOEXEC) = 0
[pid 1849389] 21:45:37 fcntl(60, F_SETFD, FD_CLOEXEC) = 0
<<< Child closes the read part of the coproc -> parent pipe >>>
[pid 1849390] 21:45:37 close(63)        = 0
<<< Child closes the write part of the parent -> coproc pipe >>>
[pid 1849390] 21:45:37 close(60)        = 0
<<< Child rebinds the pipe to stdin (FD 0) and stdout (FD 1)
[pid 1849390] 21:45:37 dup2(61, 0)      = 0
[pid 1849390] 21:45:37 close(61)        = 0
[pid 1849390] 21:45:37 dup2(62, 1)      = 1
[pid 1849390] 21:45:37 close(62)        = 0
<<< Child starts reading from the pipe >>>
[pid 1849390] 21:45:37 read(0,  <unfinished ...>
<<< Parent launches the `sleep 1` command >>>
[pid 1849389] 21:45:37 clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f72dd92fa10) = 1849391
strace: Process 1849391 attached
[pid 1849391] 21:45:38 +++ exited with 0 +++
[pid 1849389] 21:45:38 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=1849391, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
<<< Sleep exits after 1 sec >>>
[pid 1849389] 21:45:38 write(1, "foo\n", 4) = 4

<<< Parent will now exec `sleep 2` >>>
<<< Child read now reads EOF >>>
[pid 1849390] 21:45:38 <... read resumed>"", 1) = 0
[pid 1849390] 21:45:38 write(1, "bar\n", 4) = -1 EPIPE (Broken pipe)
[pid 1849390] 21:45:38 --- SIGPIPE {si_signo=SIGPIPE, si_code=SI_USER, si_pid=1849390, si_uid=1000} ---
[pid 1849389] 21:45:38 close(3)         = 0
[pid 1849390] 21:45:38 +++ killed by SIGPIPE +++
21:45:38 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_KILLED, si_pid=1849390, si_uid=1000, si_status=SIGPIPE, si_utime=0, si_stime=0} ---
21:45:40 +++ exited with 0 +++
```

</details>

Reply:
Can we avoid the sleep?

Reply:
(If I'm not wrong, on hydra.nixos.org, this has the potential to exercise spurious failures under heavy load…)

Reply:
Found a way, using `fdflags`.